rfc/rfc-40/rfc-40.md

```diff
@@ -0,0 +1,195 @@
+we plan to reuse the `ParquetPageSource` in the Hive connector. This has the advantage of using Trino's
+custom `ParquetReader` that can efficiently skip data sections by using statistics in file headers/footers. This is also
+where we will handle the column projections and build predicates for the parquet reader.
+
+```java
+public class HudiPageSourceProvider
+    implements ConnectorPageSourceProvider {
+  private final HdfsEnvironment hdfsEnvironment;
+  private final FileFormatDataSourceStats fileFormatDataSourceStats;
+  private final ParquetReaderOptions parquetReaderOptions;
+  private final DateTimeZone timeZone;
+}
+```
+In summary, Trino coordinator uses the metadata and split manager APIs to gather information about the table and partitions to
+generate a query plan and logical splits of the table contents. Each split is processed by a task in the Trino worker.
+Here, workers invoke the page source APIs as tasks produce data in the form of pages.
+Subsequently, native (parquet) reader read the block of pages while executing the query.
+
+### Improving split parallelism
+
+In order to improve split parallelism, we will push the fetch of partitions down to split source.
+Since each partition is independent of another, so we batch them up and each batch is processed by a thread.
+Additionally, the full list of partition names are fetched just once from the metastore and the subsequent
+construction of the partition key from the partition name can be done in parallel as well.
+
+### Improving listing
+In order to improve listing, we assume that the path exists,
+and so we bypass the `FileSystem#exists` check in `AbstractHoodieTableFileSystemView` while fetching latest base files.
+The connector will also support metadata-based listing which will retrieve partition listings from Hudi's internal metadata table. 
+This should further help improve the performance.
+
+### Snapshot queries on MOR table
+
+This requires merging base file and log files.
+One way is to use the `HoodieRealtimeRecordReader` which can do compacted reading.
+However, this means we will have to give up Trino's optimized parquet reader.
+Another way is to enumerate the merged splits and use the native reader.
+This can be done in `HoodieRealtimeInputFormatUtils#getRealtimeSplits()` which is invoked in `HoodieParquetRealtimeInputFormat`.
+We can reuse this logic for reading MOR table via the connector.
+
+## Rollout/Adoption Plan
+
+- What impact (if any) will there be on existing users?
+
+There will be no impact on existing users because this is a new connector. It does not change the behavior of current
+integration through the existing Hive connector. It gives users more choice.
+
+- What do we lose if we move away from the Hive connector?
+
+Hive connector takes advantage of [caching](https://trino.io/docs/current/connector/hive-caching.html) to reduce load on
```

Comment:
could we capture these as JIRAs on the epic?

Reply:
Yes. Will add a link in the rfc as well.
HUDI-3339